README.md

[![minimal R
version](https://img.shields.io/badge/R%3E%3D-3.5.0-6666ff.svg)](https://cran.r-project.org/)
[![GitHub
(pre-)release](https://img.shields.io/github/release/ropensci/rrricanes/all.svg)](https://github.com/ropensci/rrricanes/tags)
[![](https://badges.ropensci.org/118_status.svg)](https://github.com/ropensci/onboarding/issues/118)
[![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/rrricanes)](https://cran.r-project.org/package=rrricanes)
[![Travis Build
Status](https://img.shields.io/travis/ropensci/rrricanes/master.svg)](https://travis-ci.org/ropensci/rrricanes)
[![AppVeyor Build
Status](https://img.shields.io/appveyor/ci/timtrice/rrricanes-g4dos/master.svg)](https://ci.appveyor.com/project/timtrice/rrricanes-g4dos)
[![Docker Build
Status](https://img.shields.io/docker/cloud/build/timtrice/rrricanes.svg?style=popout)](https://cloud.docker.com/repository/docker/timtrice/rrricanes)
[![codecov](https://codecov.io/gh/ropensci/rrricanes/branch/master/graph/badge.svg)](https://codecov.io/gh/ropensci/rrricanes)
[![Launch
rstudio.cloud](https://img.shields.io/badge/launch-rstudio.cloud-yellowgreen.svg)](https://rstudio.cloud/project/400381)

# rrricanes (0.2.0.6.9002) <img src='man/figures/logo.png' align="right" height="138" />

## Web scraper for Atlantic and east Pacific hurricanes and tropical storms

`rrricanes` is a R library that extracts information from [available
archives](http://www.nhc.noaa.gov/archive/1998/1998archive.shtml) on
past and current tropical cyclones. Currently, archives date back to
1998.

Data can be obtained for cyclones in the north Atlantic (considered the
Atlantic Basin) and north-eastern Pacific (the East Pacific Basin from
140°W and eastward.

Central Pacific data (140°W to 180°W) is included if issued by the
[National Hurricane Center](http://www.nhc.noaa.gov/) (generally they’re
issued by the [Central Pacific Hurricane
Center](http://www.prh.noaa.gov/cphc/)).

This library parses the text advisories of all tropical cyclones since
1998. Over the years the formats of the text products have changed and
many are loosely formatted.

I wrote this package with the goal of consolidating messy text data into
well-organized formats that can easily be saved to CSV, SQL and other
data formats.

### Advisory Products

Generally speaking, there are five products available for tropical
cyclones issued at 03:00, 09:00, 15:00 and 21:00 UTC;

1.  Storm Discussion - These are technical discussions centered on the
    current structure of the cyclone, satellite presentation, computer
    forecast model tendencies and more.

2.  Forecast/Adivsory - This data-rich product lists the current
    location of the cyclone, its wind structure, forecast and forecast
    wind structure.

3.  Public Advisory - These are general text statements issued for the
    public-at-large. Information in these products is a summary of the
    Forecast/Advisory product along with any watches and warnings
    issued, changed, or cancelled. Public Advisory products are the only
    regularly-scheduled product that may be issued intermittently (every
    three hours and, occasionally, every two hours) when watches and
    warnings are in effect.

4.  Wind Speed Probabilities - These products list the probability of a
    minimum sustained wind speed expected in a given forecast window.
    This product replaces the Strike Probabilities product beginning in
    2006 (see below).

5.  Updates - Tropical Cyclone Updates may be issued at any time if a
    storm is an immediate threat to land or if the cyclone undergoes a
    significant change of strength or structure. The information in this
    product is general.

**Discontinued Products**

These products are included in the package though they have been
discontinued at some point:

1.  Strike Probabilities - List the probability of a tropical cyclone
    passing within 65 nautical miles of a location within a forecast
    window. Replaced in 2006 by the Wind Speed Probabilities product.

2.  Position Estimates - Typically issued as a storm is threatening land
    but generally rare (see Hurricane Ike 2008, Key AL092008). It is
    generally just an update of the current location of the cyclone.
    After the 2011 hurricane season, this product was discontinued;
    Updates are now issued in their place.

### Getting Started

Please view the vignette ‘Getting Started’:

``` r
vignette("getting_started", package = "rrricanes")
```

[Online documentation](https://timtrice.github.io/rrricanes/) is also
available.

#### Prerequisites

`rrricanes` requires an active internet connection as data is extracted
from online sources.

Linux users must also have the `libgdal-dev`, `libproj-dev` and
`libxml2-dev` packages installed.

To add `rrricanesdata`, a [package of post-scraped
datasets](https://github.com/ropensci/rrricanesdata),

``` r
remotes::install_github("ropensci/rrricanesdata")
```

To use high resolution tracking maps you will need to install the
`rnaturalearthhires` package.

``` r
remotes::install_github("ropensci/rnaturalearthhires")
```

#### Installing

`rrricanes` is currently only available in GitHub. It can be installed
using the `devtools` package:

``` r
devtools::install_github("ropensci/rrricanes", build_vignettes = TRUE)
```

### Built With

  - [R 3.5.0](https://www.r-project.org/) - The R Project for
    Statistical Computing

### Contributing

Please read
[CONTRIBUTING.md](https://github.com/ropensci/rrricanes/blob/master/.github/CONTRIBUTING.md)
for details on our code of conduct, and the process for submitting pull
requests to us.

### Versioning

We use [SemVer](http://semver.org/) for versioning. For the versions
available, see the [tags on this
repository](https://github.com/ropensci/rrricanes/tags).

### Authors

  - **Tim Trice** - *Initial work* -
    [timtrice](https://github.com/timtrice)

See also the list of
[contributors](https://github.com/ropensci/rrricanes/contributors) who
participated in this project.

### License

This project is licensed under the MIT License - see the
[LICENSE.md](LICENSE.md) file for details

### Acknowledgments

  - [Molyneux, James](https://github.com/jimmylovestea)
  - [Padgham, Mark](https://github.com/mpadge)
  - [Robinson, Emily](https://github.com/robinsones)
  - [Rudis, Bob](https://github.com/hrbrmstr)
  - [Salmon, Maëlle](https://github.com/maelle)
  - [Stachelek, Joseph](https://github.com/jsta)

### Known Data Quality Issues

1.  Hurricane Juan (AL152003), Adv 15; no status leads to improper
    `Status` and `Name` values in some datasets.
    ([\#82](https://github.com/ropensci/rrricanes/issues/82))

<center>

[![](http://ropensci.org/public_images/github_footer.png)](http://ropensci.org)

</center>


NEWS.md
rrricanes Unreleased (yyyy-mm-dd)
==================================

### NEW FEATURES

* Add `get_storm_list` that returns a list of, it would appear, all known storms for many basins (beyond AL, EP)
  + Currently only returns 2582 cyclones despite more storms present in the file. This is due to a currenty unknown issue in the raw CSV file.

* `get_ftp_storm_data`, given a `stormid` (key) and a product, will return a dataframe of the requested data if it exists for the current storm. This opens up cyclones earlier than 1998 **however** is not yet tested.

### MINOR IMPROVEMENTS

* Add `nm_to_sm` to convert nautical miles to survey miles. (#99)

* Allow parameters for rgdal::readOGR (#104)

* Add `tidy_adv` to replace `tidy_fstadv`; `tidy_fstadv` will be removed in release 0.2.2 (#103)

* `al_prblty_stations` and `cp_prblty_stations` updated 
  to accomodate new dataset. (#46)
  
* `ep_prblty_stations` now returns dataset of east Pacific stations. (#46)

* `tracking_chart` and associated functions have been removed in favor of rnaturalearth and related packages. See `vignette("gis_data", package = "rrricanes") for details.

* `gis_download` no longer reads in the shapefiles; only downloads and extracts the zip files.

* `shp_to_df` also removed in favor of built-in `sf` and `ggplot2` functions.

### BUG FIXES

* `tidy_fcst` only returned forecasts for hours 12:96. Now returns 120. (#107)
* `gis_wsp` now uses GET instead of POST to retrieve wind speed probability GIS datasets. (#108)

### DEPRECATED AND DEFUNCT

* NA

rrricanes 0.2.0-6 (2017-07-16)
==================================

### NEW FEATURES

* NA

### MINOR IMPROVEMENTS

* `get_storms` and `get_storm_data` have been rewritten to utilize pkg `crul`'s asynchronous features. This will not make much of a difference in `get_storms` (and may actually be slightly slower; to be explained). But the difference with `get_storm_data` should be very noticeable. There is a limit to hitting the NHC archives; 80 requests per 10 seconds. Both functions send 4 links through every 0.5 seconds to avoid this limit. Timeout issues should no longer occur so options rrricanes.http_attempts and rrricanes.http_timeout have been removed. The primary cause of long processing now is due to scraping, particularly with the `fstadv` products; the amount of data in these products and the unstructured nature of the products require a number of rules. This can probably be simplified in future releases. (#94)

### BUG FIXES

* NA

### DEPRECATED AND DEFUNCT

* `load_storm_data` has been removed in favor of loading `rrricanesdata`. See 
  vignette "installing_rrricanesdata" for more information. 

rrricanes 0.2.0-5 (2017-07-08)
==================================

### NEW FEATURES

* NA

### MINOR IMPROVEMENTS

* `load_storm_data` now takes `readr::read_csv` parameters.

### BUG FIXES

* NA

### DEPRECATED AND DEFUNCT

* NA

rrricanes 0.2.0-4 (2017-06-25)
==================================

### NEW FEATURES

* NA

### MINOR IMPROVEMENTS

* `Key` variable added to `discus` dataframes. `Key` will be NA for all cyclones >= 2005. Should not be <= 2006. (#80)
* Removed `Adv` variable from `posest` dataframes. Position estimates do not have advisory numbers. (#81)
* Removed `Adv` variable from `update`. Updates do not have advisory numbers. (#84)
* Added variable `Key` to `get_public` dataframes. (#85)
* Added variable `Key` to `get_update` dataframes. (#86)
* Removed non-existent wind radii variables in `get_fstadv`. Hrs 48 and 72 hours only have 34 and 50kt wind fields. Hrs 96 and 120 have none. (#89)

### BUG FIXES

* NA

### DEPRECATED AND DEFUNCT

* NA

rrricanes 0.2.0-3 (2017-06-22)
==================================

### NEW FEATURES

* NA

### MINOR IMPROVEMENTS

* Examples for functions `knots_to_mph`, `mb_to_in`, `status_abbr_to_str`, `get_discus`, `get_fstadv`, `tidy_fstadv`, `tidy_wr`, `tidy_fcst` and `tidy_fcst_wr`.
* Minor documentation updates
* Small enhancements to tests, vignettes

### BUG FIXES

* NA

### DEPRECATED AND DEFUNCT

* NA

rrricanes 0.2.0-2 (2017-06-22)
==================================

### NEW FEATURES

* NA

### MINOR IMPROVEMENTS

* NA

### BUG FIXES

* Advisories now issued when tropical cyclone development is anticipated, but not yet occurred, and watches and warnings need to be issued. See AL022017, AL032017. 
* Added additional time zones (HDT, HST, MDT, MST)
* Appended additional headers (TCE, WTPA, MIATCM, MIAWRKAD1, MIAWRKAP)

### DEPRECATED AND DEFUNCT

* NA

rrricanes 0.2.0-1 (2017-06-16)
==================================

### NEW FEATURES

* GIS functions now available. Please note some of these products may not exist for every available cyclone/advisory.
 + `gis_advisory` Typically will include current and past track data, forecast track data, forecast cone (margin of error) and wind radius data. 
 + `gis_breakpoints` List of breakpoints typically used for watch/warning areas but is not a requirement.
 + `gis_latest` Retrieves the latest GIS products for all active storms. 
 + `gis_outlook` Retrives the latest tropical weather outlook in shapefile format.
 + `gis_prob_storm_surge` Probabilistic storm surge; a polygon dataset for psurge and esurge products with various criteria.
 + `gis_windfield` Wind radius datasets.
 + `gis_wsp` Wind speed probabilities
 + `gis_download` Use this function to download the URLs returned from the above functions.
 + `shp_to_df` added to convert lines and polygons spatial dataframes to dataframes. Points dataframes can be converted using `tibble::as_dataframe` (target the @data object).

### MINOR IMPROVEMENTS

* [Enhanced documentation](https://ropensci.github.io/rrricanes/) added online using `pkgdown`. 

R/base.R
#' @title rrricanes
#' @description rrricanes is a web-scraping library for R designed to deliver
#' hurricane data (past and current) into well-organized datasets. With these
#' datasets you can explore past hurricane tracks, forecasts and structure
#' elements.
#'
#' This documentation and additional help articles
#' \href{https://ropensci.github.io/rrricanes/}{can be found online}.
#'
#' Text products (Forecast/Advisory, Public Advisory, Discussions and
#' Probabilities) are only available from 1998 to current. An effort will be
#' made to add prior data as available.
#'
#' @section Getting Storms:
#' List all storms that have developed by year and basin. Year must be in a
#' four-digit format (\%Y) and no earlier than 1998. Basin can be one or both
#' of Atlantic ("AL") or East Pacific ("EP").
#' \describe{
#'   \item{\code{\link{get_storms}}}{List all storms by year, basin}
#' }
#'
#' @section Getting Storm Data:
#'
#' \code{\link{get_storm_data}} can be used to select multiple products,
#' multiple storms and from multiple basins.
#'
#' Additional text products are:
#' \describe{
#'   \item{\code{\link{get_discus}}}{Storm Discussions}
#'   \item{\code{\link{get_fstadv}}}{Forecast/Advisory. These products contain a
#'   bulk of the information for tropical cyclones including current position,
#'   structure, forecast position and forecast structure.}
#'   \item{\code{\link{get_posest}}}{Position Estimates. Rare and used generally
#'   for threatening cyclones. This product was discontinued after the 2013
#'   season and is now issued as \code{\link{get_update}}.}
#'   \item{\code{\link{get_prblty}}}{Strike Probabilities. Show the probability
#'   of the center of a cyclone passing within 65nm of a location for a given
#'   forecast period. This product was discontinued after 2005, replaced with
#'   \code{\link{get_wndprb}}.}
#'   \item{\code{\link{get_public}}}{Public Advisory. General non-structured
#'   information exists in these products.}
#'   \item{\code{\link{get_update}}}{Updates. Generally issued when a cyclone
#'   undergoes a sudden change that requires immediate notice.}
#'   \item{\code{\link{get_wndprb}}}{Wind Speed Probability. Lists the
#'   probability of a location experiencing a minimum of 35kt, 50kt or 64kt
#'   winds for an alotted forecast period or accumulated probability. This
#'   product replaced \code{\link{get_prblty}} after the 2005 season.}
#' }
#'
#' The products above may take some time to load if the NHC website is slow (as
#' is often the case, unfortunately). For all storm advisories issued outside
#' of the current month, use the \code{rrricanesdata} package.
#'
#' To install \code{rrricanesdata}, run
#'
#' \code{
#' remotes::install_github("ropensci/rrricanesdata")
#' }
#'
#' See \code{vignette("installing_rrricanesdata", package = "rrricanes")} for
#' more information.
#'
#' @section GIS Data:
#'
#' For enhanced plotting of storm data, several GIS datasets are available. The
#' core GIS functions return URLs to help you refine the data you wish to view.
#' (Some products will not exist for all storms/advisories). These products are:
#'
#' \describe{
#'   \item{\code{\link{gis_advisory}}}{Past track, current position, forecast
#'     and wind radii}
#'   \item{\code{\link{gis_breakpoints}}}{Breakpoints for watches and warnings}
#'   \item{\code{\link{gis_latest}}}{All available GIS products for active
#'     cyclones}
#'   \item{\code{\link{gis_outlook}}}{Tropical Weather Outlook}
#'   \item{\code{\link{gis_prob_storm_surge}}}{Probabilistic Storm Surge}
#'   \item{\code{\link{gis_windfield}}}{Wind Radii}
#'   \item{\code{\link{gis_wsp}}}{Wind Speed Probabilities}
#' }
#'
#' \code{\link{gis_download}} will download the datasets from the above
#' functions.
#'
#' @section Package Options:
#'
#' \code{dplyr.show_progress} displays the dplyr progress bar when scraping raw
#' product datasets. In \code{\link{get_storms}}, it is based on the number of
#' years being requested. In the product functions (i.e.,
#' \code{\link{get_fstadv}}) it is based on the number of advisories. It can be
#' misleading when calling \code{\link{get_storm_data}} because it shows the
#' progress of working through a storm's product advisories but will reset on
#' new products/storms.
#'
#' @section Package Options:
#'
#' \code{dplyr.show_progress} displays the dplyr progress bar when scraping raw
#' product datasets. In \code{\link{get_storms}}, it is based on the number of
#' years being requested. In the product functions (i.e.,
#' \code{\link{get_fstadv}}) it is based on the number of advisories. It can be
#' misleading when calling \code{\link{get_storm_data}} because it shows the
#' progress of working through a storm's product advisories but will reset on
#' new products/storms.
#'
#' \code{rrricanes.working_msg} is set to FALSE by default. When TRUE, it will
#' list the current storm, advisory and date being worked.
#'
#' @docType package
#' @name rrricanes
NULL

#' @importFrom magrittr %>%
#' @importFrom rlang .data

.pkgenv <- new.env(parent = emptyenv())

.onLoad <- function(libname, pkgname) {
  op <- options()
  op.rrricanes <- list(rrricanes.working_msg = FALSE)
  toset <- !(names(op.rrricanes) %in% names(op))
  if (any(toset)) options(op.rrricanes[toset])
  invisible()
  has_data <- base::requireNamespace("rrricanesdata", quietly = TRUE)
  .pkgenv[["has_data"]] <- has_data
}

.onAttach <- function(libname, pkgname) {
  msg <- "rrricanes is not intended for use in emergency situations."
  packageStartupMessage(msg)
}

hasData <- function(has_data = .pkgenv$has_data) {
  if (!has_data) {
    stop("rrricanesdata is not installed.")
  }
}

utils::globalVariables(c(
  "Date", "Hour", "Minute", "Lat", "LatHemi", "Lon",
  "LonHemi", "Wind", "Gust", "Month", "Year", "FcstDate",
  "WindField34", "WindField50", "WindField64", "lat",
  "long", "group", ".", "NW34", "name", "data", "Basin",
  stringr::str_c(c("NE", "SE", "SW", "NW", "64")),
  stringr::str_c(c("NE", "SE", "SW", "NW", "50")),
  stringr::str_c(c("NE", "SE", "SW", "NW", "34"))
))

#' @title extract_year_archive_link
#' @description Extracts the year from the archive link.
#' @param link URL of archive page
#' @return year 4-digit numeric
#' @keywords internal
extract_year_archive_link <- function(link) {
  # Year is listed in link towards the end surrounded by slashes.
  as.numeric(stringr::str_match(link, "/([:digit:]{4})/")[, 2])
}

#' @title get_url_contents
#' @description Get contents from URL
#' @details This function primarily is reserved for extracting the contents of
#' the individual products \(thought it can be used in other instances\). Often,
#' there are timeout issues. This is an attempt to try to work around that.
#' @param link URL to download
#' @keywords internal
get_url_contents <- function(links) {
  download_text <- function(grouped_links) {

    # Create a new Async object with `grouped_links`
    grouped_links <- crul::Async$new(urls = grouped_links)

    # Get `grouped_links`
    results <- grouped_links$get()

    # Do we have any bad `grouped_links`?
    bad_results_ind <- which(purrr::map(results, ~ .$success()) == FALSE)
    if (length(bad_results_ind) > 0) {
      warning(sprintf(
        "URL %s was unsuccesful.\n",
        purrr::map(results[bad_results_ind], ~ .$url)
      ),
      call. = FALSE
      )
      # Remove bad `grouped_links`
      results <- results[-bad_results_ind]
    }
    purrr::map_chr(results, ~ .$parse("UTF-8"))
  }

  # Create groups of links divisible by 80. We are to allow no more than 80
  # requests every 10 seconds. If length of `link` is less than 80, then will
  # only have one group and should have no delay.
  groups <- ceiling(seq_along(1:length(links)) / 80)
  links <- split(links, groups)

  # Set progress bar
  p <- dplyr::progress_estimated(n = length(links))

  contents <-
    links %>%
    purrr::imap(.f = function(x, y) {
      if (as.numeric(y) != length(links)) {
        # Send group of links to `download_txt`
        txt <- download_text(x)
        # We are not in the last group; apply a delay
        p$tick()$print()
        if (getOption("rrricanes.working_msg")) {
          message("Waiting 10 seconds to retrieve large numbers of links.")
        }
        p$pause(10)
        txt
      } else {
        # Send group of links to `download_txt`
        p$tick()$print()
        download_text(x)
      }
    })

  purrr::flatten_chr(contents)
}

#' @title get_nhc_link
#' @description Return root link of NHC archive pages.
#' @param withTrailingSlash True, by default. False returns URL without
#' trailing slash.
#' @param protocol https or http
#' @keywords internal
get_nhc_link <- function(withTrailingSlash = TRUE, protocol = "https") {
  if (withTrailingSlash) {
    return(sprintf("%s://www.nhc.noaa.gov/", protocol))
  }
  sprintf("%s://www.nhc.noaa.gov", protocol)
}

#' @title get_nhc_ftp_link
#' @description Return root of NHC FTP server
#' @inheritParams get_nhc_link
#' @keywords internal
get_nhc_ftp_link <- function(withTrailingSlash = TRUE) {
  if (withTrailingSlash) {
    return("ftp://ftp.nhc.noaa.gov/")
  }
  "ftp://ftp.nhc.noaa.gov"
}

#' @title knots_to_mph
#' @description convert knots (kt) to miles per hour (mph)
#' @param x wind speed in knots
#' @return x in miles per hour
#' @examples
#' knots_to_mph(65)
#' @export
knots_to_mph <- function(x) {
  x * 1.15077945
}

#' @title mb_to_in
#' @description convert millibars (mb) to inches of mercury (in)
#' @param x barometric pressure in mb
#' @return x in inches
#' @examples
#' mb_to_in(999)
#' @export
mb_to_in <- function(x) {
  x * 0.029529983071
}

#' @title month_str_to_num
#' @description Convert three-character month abbreviation to integer
#' @param m Month abbreviated (SEP, OCT, etc.)
#' @return numeric 1-12
#' @keywords internal
month_str_to_num <- function(m) {
  abbr <- which(month.abb == stringr::str_to_title(m))
  if (purrr::is_empty(abbr)) {
    stop(sprintf("%s is not a valid month abbreviation.", m))
  }
  abbr
}

#' @title nm_to_sm
#' @description Convert nautical miles to survey miles
#' @param x Nautical miles
#' @examples
#' nm_to_sm(c(50, 100, 150))
#' @export
nm_to_sm <- function(x) {
  x * 1.15078
}

#' @title saffir
#' @description Return category of tropical cyclone based on wind. Saffir-
#' Simpson Hurricane Scale does not apply to non-tropical cyclones.
#' @param x Vector of wind speed values.
#' @examples
#' saffir(c(32, 45, 70, 90, 110, 125, 140))
#' @export
saffir <- function(x) {
  y <- character(length = length(x))
  y[x <= 33] <- "TD"
  y[dplyr::between(x, 34, 64)] <- "TS"
  y[dplyr::between(x, 65, 83)] <- "HU1"
  y[dplyr::between(x, 84, 95)] <- "HU2"
  y[dplyr::between(x, 96, 113)] <- "HU3"
  y[dplyr::between(x, 114, 134)] <- "HU4"
  y[x >= 135] <- "HU5"
  y
}

#' @title status_abbr_to_str
#' @description Convert Status abbreviation to string
#' @param x character vector of status abbreviations
#' @details Status abbreviations
#' \describe{
#'   \item{DB}{Disturbance (of any intensity)}
#'   \item{EX}{Extratropical cyclone (of any intensity)}
#'   \item{HU}{Tropical cyclone of hurricane intensity (> 64 knots)}
#'   \item{LO}{A low that is neither a tropical cyclone, a subtropical
#'         cyclone, nor an extratropical cyclone (of any intensity)}
#'   \item{SD}{Subtropical cyclone of subtropical depression intensity
#'         (< 34 knots)}
#'   \item{SS}{Subtropical cyclone of subtropical storm intensity
#'         (> 34 knots)}
#'   \item{TD}{Tropical cyclone of tropical depression intensity (< 34 knots)}
#'   \item{TS}{Tropical cyclone of tropical storm intensity (34-63 knots)}
#'   \item{WV}{Tropical Wave (of any intensity)}
#' }
#' @return character vector of strings
#' @seealso \url{http://www.aoml.noaa.gov/hrd/hurdat/newhurdat-format.pdf}
#' @examples
#' # Extratropical Cyclone
#' status_abbr_to_str("EX")
#'
#' # Hurricane
#' status_abbr_to_str("HU")
#' @export
status_abbr_to_str <- function(x) {
  y <- character(length = 0)
  y[x == "TD"] <- "Tropical Depression"
  y[x == "TS"] <- "Tropical Storm"
  y[x == "HU"] <- "Hurricane"
  y[x == "EX"] <- "Extratropical Cyclone"
  y[x == "SD"] <- "Subtropical Depression"
  y[x == "SS"] <- "Subtropical Storm"
  y[x == "LO"] <- "Low"
  y[x == "WV"] <- "Tropical Wave"
  y[x == "DB"] <- "Disturbance"
  y
}
